Feature/jfb 3 ticket hierarchy

Arthur2224 opened the pull request:
**Summary of Changes**

-     Interface Implementation: Added Identifiable and Printable interfaces for standardized ID management and print functionality.

-     Ticket Attributes: Implemented getters and setters for time and stadiumSectors in Ticket to allow controlled access and modification.

-     Package Structure: Organized code into packages:

-         entities

-         enums

-         utilities

-     ShareableTicket Class: Created ShareableTicket, extending Ticket

-     User Hierarchy: Introduced an abstract User class, extended by Client and Admin 

-     Equality and Hashing in Ticket: Overridden equals() and hashCode() in Ticket

Arthur2224 committed:
Implemented interfaces Identifiable(ability to implement id) and Printable (ability to implement .print() method)

Arthur2224 committed:
Implemented getters and setters for 'time' and 'StadiumSectors' values.

Arthur2224 committed:
Create package structure for entities, enums, and utilities; move files accordingly.

Arthur2224 committed:
Implemented ShareableTicket class, updated setTime method in Ticket to set current time, and added usage example in TicketService. Increased usage of setTime method

Arthur2224 committed:
Implemented abstract class User and extended it with Client and Admin classes.

Arthur2224 committed:
Overridden equals() and hashCode() methods in Ticket class

Arthur2224 committed:
Implemented @Nullable annotation

Darya-Drutsko commented on src/main/java/org/example/annotations/AnnotaionProcessor.java:
```diff
@@ -0,0 +1,27 @@
+import java.lang.reflect.Field;
+
+public class AnnotaionProcessor {
+    public static void checkNullValues(Object object) { // Can be used for any class
```
Cool!

Darya-Drutsko commented on src/main/java/org/example/interfaces/Identifiable.java:
```diff
@@ -0,0 +1,6 @@
+package org.example.interfaces;
+
+public interface Identifiable {
+    int getId();
```
it's seems like getter and setter - probably we need also a field - wdyt?

Arthur2224 replied:
yes, it is better to choose an abstract class as a solution to the issue because there is no need to modify methods after initialization. Also, abstract classes can contain variables without obligatory initialization unlike interfaces

Darya-Drutsko commented on src/main/java/org/example/interfaces/Printable.java:
```diff
@@ -0,0 +1,7 @@
+package org.example.interfaces;
+
+public interface Printable {
```
good choice!

Darya-Drutsko commented on src/main/java/org/example/entities/Admin.java:
```diff
@@ -0,0 +1,11 @@
+    @Override
+    public void printRole(){
+        System.out.println("ROLE: ADMIN");
+    }
```
formatting - add empty row after method ;)

Darya-Drutsko commented on src/main/java/org/example/entities/Client.java:
```diff
@@ -0,0 +1,11 @@
+    @Override
+    public  void printRole(){
+        System.out.println("Role: CLIENT");
+    }
```
formatting - add empty row after method ;)
and remove empty spaces in method signature

Darya-Drutsko commented on src/main/java/org/example/entities/ShareableTicket.java:
```diff
@@ -0,0 +1,18 @@
+
+public class ShareableTicket extends Ticket{
+    public ShareableTicket(StadiumSectors stadiumSector) {
+        super();
```
why do we need it here? In which cases we use super()?

Arthur2224 replied:
empty parent constructor called automaticaly, there are no need to call it :)
It needs to be called when the parent class does't contain an empty constructor, or when you need to call a specific constructor. Also, but not related to this case, super() can be used to call methods overridden from the parent class (to access the original versions)

Darya-Drutsko commented on src/main/java/org/example/TicketService.java:
```diff
@@ -25,22 +30,51 @@ public static void main(String[] args) {
-        System.out.println( "Find the ticked by ID ["+ticket_1.getId()+"]: " + ticketManager.findTicketById(ticket_1.getId()));
+        System.out.println("Overwriting equal, hashcode:");
 
+        System.out.println(ticket_1.equals(new Admin())); // false
```
usually we don't use _ or numbers in variable naming :)

Darya-Drutsko commented on src/main/java/org/example/entities/Ticket.java:
```diff
@@ -1,13 +1,18 @@
-    private String id;
+public class Ticket implements Identifiable, Printable {
+    @Nullable(key = "ID must not be NULL")
+    private Integer id; // switched to Integer because of default value of this type is Null for annotation task
```
related to comment - and what is default value for String? :)

Arthur2224 replied:
also Null

Darya-Drutsko commented on src/main/java/org/example/entities/Ticket.java:
```diff
@@ -1,13 +1,18 @@
+public class Ticket implements Identifiable, Printable {
+    @Nullable(key = "ID must not be NULL")
+    private Integer id; // switched to Integer because of default value of this type is Null for annotation task
+    private long time;
```
discussed during lesson - use, please, specific classes for time :)

Darya-Drutsko commented on src/main/java/org/example/utilities/TicketUtility.java:
```diff
@@ -1,9 +1,12 @@
 import java.util.List;
 
-public class TicketManager {
+public class TicketUtility {
```
better to use manager in naming
when we're talking about utility classes we usually use there static methods without object creation

Darya-Drutsko reviewed (commented):
Good job!
left few comments and overall recommendation don't forget about formatting

Arthur2224 committed:
Formatting:
- Removed super() from ShareableTicket constructor
- Rejected use of '_' and numbers in variable names
- Switched Time variable to LocalDateTime

Arthur2224 committed:
Formatting:
- Renamed TicketUtility to TicketManager
- Refactored ID handling in Ticket class: switched from Interface to Abstract class
